src/System.Linq/tests/WhereTests.cs

```diff
@@ -0,0 +1,644 @@
+            
+            Assert.Equal(2, result.Count());
+            Assert.Equal(2, result.ElementAt(0));
+            Assert.Equal(4, result.ElementAt(1));
+        }
+
+        [Fact]
+        public void WhereWhere_IReadOnlyCollection_ReturnsExpectedValues()
+        {
+            IReadOnlyCollection<int> source = new ReadOnlyCollection<int>(new List<int> { 1, 2, 3, 4, 5 });
+            Func<int, bool> evenPredicate = (value) => value % 2 == 0;
+
+            IEnumerable<int> result = source.Where(evenPredicate).Where(evenPredicate);
+
+            Assert.Equal(2, result.Count());
+            Assert.Equal(2, result.ElementAt(0));
+            Assert.Equal(4, result.ElementAt(1));
+        }
+
+        [Fact]
+        public void WhereWhere_ICollection_ReturnsExpectedValues()
+        {
+            ICollection<int> source = new LinkedList<int>(new List<int> { 1, 2, 3, 4, 5 });
+            Func<int, bool> evenPredicate = (value) => value % 2 == 0;
+
+            IEnumerable<int> result = source.Where(evenPredicate).Where(evenPredicate);
+
+            Assert.Equal(2, result.Count());
+            Assert.Equal(2, result.ElementAt(0));
+            Assert.Equal(4, result.ElementAt(1));
+        }
+
+        [Fact]
+        public void WhereWhere_IEnumerable_ReturnsExpectedValues()
+        {
+            IEnumerable<int> source = Enumerable.Range(1, 5);
+            Func<int, bool> evenPredicate = (value) => value % 2 == 0;
+
+            IEnumerable<int> result = source.Where(evenPredicate).Where(evenPredicate);
+
+            Assert.Equal(2, result.Count());
+            Assert.Equal(2, result.ElementAt(0));
+            Assert.Equal(4, result.ElementAt(1));
+        }
+
+        #endregion
+
+        #region Exceptions
+        
+        [Fact]
+        public void Where_PredicateThrowsException()
+        {
+            int[] source = new[] { 1, 2, 3, 4, 5 };
+            Func<int, bool> predicate = value =>
+            {
+                if (value == 1)
+                {
+                    throw new InvalidOperationException();
+                }
+                return true;
+            };
+
+            var result = source.Where(predicate);
+            var enumerator = result.GetEnumerator();
+
+            // Ensure the first MoveNext call throws an exception
+            Assert.Throws<InvalidOperationException>(() => enumerator.MoveNext());
+
+            // Ensure Current is set to the default value of type T
+            int currentValue = enumerator.Current;
+            Assert.Equal(default(int), currentValue);
+
+            // Ensure subsequent MoveNext calls succeed
+            Assert.True(enumerator.MoveNext());
+            Assert.Equal(2, enumerator.Current);
+        }
+
+        /// <summary>
+        /// Test enumerator - returns int values from 1 to 5 inclusive.
+        /// </summary>
+        private class TestEnumerator : IEnumerable<int>, IEnumerator<int>
```

Comment:
Can this be factored out into a project-wide helper?  Looks the same as the one used in the Select tests.